cmd/estuary-shuttle/ppm.go

```diff
@@ -0,0 +1,162 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"sort"
+	"time"
+
+	apiv2 "github.com/application-research/estuary/api/v2"
+	"github.com/application-research/estuary/node"
+	util "github.com/application-research/estuary/util"
+	"github.com/filecoin-project/go-address"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+	"github.com/libp2p/go-libp2p/p2p/protocol/ping"
+	"github.com/multiformats/go-multiaddr"
+)
+
+type PeerPingManager struct {
+	Node     *node.Node
+	HtClient *ShuttleHttpClient
+	Result   PingManyResult
+}
+
+type SpHost struct {
+	spAddr    address.Address
+	multiAddr []multiaddr.Multiaddr
+	peerID    peer.ID
+}
+
+type PingManyResult []PingResult
+type PingResult struct {
+	Address address.Address `json:"address"`
+	Latency int64           `json:"latency"`
+}
+
+func NewPPM(node *node.Node, shtc *ShuttleHttpClient) *PeerPingManager {
+
+	return &PeerPingManager{
+		Node:     node,
+		HtClient: shtc,
+		Result:   make(PingManyResult, 0),
+	}
+}
+
+// Kicks off a PPM task to get a list of SPs and ping them every <interval>
+// Spawns a new go thread, and returns immediately
+func (ppm *PeerPingManager) Run(interval time.Duration) {
+	go func() {
+		ctx := context.TODO()
+
+		for {
+			hosts, err := ppm.getSpList()
+			if err != nil {
+				log.Errorf("ppm could not get SP list %v", err)
+			}
+
+			ppm.PingMany(ctx, hosts)
+
+			log.Debugf("ping completed on %d hosts", len(hosts))
+
+			time.Sleep(interval)
+		}
+	}()
+}
+
+func (ppm *PeerPingManager) getSpList() ([]SpHost, error) {
+	resp, closer, err := ppm.HtClient.MakeRequest("GET", "/v2/storage-providers", nil, "")
+	if err != nil {
+		return nil, err
+	}
+	defer closer()
+
+	var out []apiv2.StorageProviderResp
+	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
+		return nil, err
+	}
+
+	var sps []SpHost
+
+	for _, m := range out {
+		ma, err := util.ToMultiAddresses(m.ChainInfo.Addresses)
+
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+		sp := SpHost{
+			spAddr:    m.Addr,
+			peerID:    m.ChainInfo.PeerID,
+			multiAddr: ma,
+		}
+		sps = append(sps, sp)
+	}
+
+	return sps, nil
+}
+
+// Ping a list of hosts, returning RTT for each of them. Errors will be ignored, unresponsive hosts will not be returned
+func (ppm *PeerPingManager) PingMany(ctx context.Context, hosts []SpHost) {
+	result := make(PingManyResult, 0)
+
+	// TODO: Batch them in go funcs for faster performance
+	for _, h := range hosts {
```

Comment:
why not do the pins in goroutines and wait for all result, so it will be faster, just like you did for the chaininfo?

Reply:
I'm hesitant to do it since the ping result is sensitive to time - we don't want a situation where it appears like an SP has bad latency when the issue is on our side, due to us running 100+ pings concurrently. I'm not sure if it could cause a queueing delay in seeing the response. Probably OK for a small number but as it scales up we might see that issue. 

I think the solution would be to only do 10 pings at a time, so the code is a little more complex. However, since we only run this routine once every 12h and it takes max. 1 second per provider, we really shouldn't need to optimize it much 🤔 

thoughts?

Reply:
thats ok then